### PR TITLE
feat(useClickOutside): introduce

### LINKED ISSUE #5348: The Dropdown Component should handle clicks outside
Currently, the Dropdown component doesn't handle clicks outside of the Component.

We should create a generic Hook, such as `useClickOutside` or something that can trigger a callback once or something of the sort.

Then we can incorporate this Hook into the Dropdown Component. This would be useful for closing the Dropdown when clicking outside of the Dropdown.

This work should be done within the `major/website-redesign` branch.

### PR DESCRIPTION
**Pull request by AugustinMauroy**
## Description

* Adding react hook `useClickOutside`
* New Pr open after some problems form **REBASE**

## Related Issues

close: [The Dropdown Component should handle clicks outside #5348](https://github.com/nodejs/nodejs.org/issues/5348)

### Check List

- [X] I have read the [Contributing Guidelines](https://github.com/nodejs/nodejs.org/blob/main/CONTRIBUTING.md) and made commit messages that follow the guideline.
- [X] I have run `npx turbo lint` to ensure the code follows the style guide. And run `npx turbo lint:fix` to fix the style errors if necessary.
- [X] I have run `npx turbo format` to ensure the code follows the style guide.
- [x] I have run `npx turbo test` to check if all tests are passing, and/or `npx turbo test:snapshot` to update snapshots if I created and/or updated React Components.
- [X] I've covered new added functionality with unit tests if necessary.

### CODE REVIEW

File: hooks/useClickOutside.ts
Comment (ovflowd) [inline hooks/useClickOutside.ts]: Can you outsource the type of this Hook?

`type UseClickOutsideProps = { ref: RefObject<HTMLElement>, handler: TypeHandler }`

`export const useClickOutside = (props: UseClickOutsideProps) => ...`

Comment (AugustinMauroy) [reply hooks/useClickOutside.ts]: Sure!

File: hooks/useClickOutside.ts
Comment (ovflowd) [inline hooks/useClickOutside.ts]:
```tsx
if (props.ref.current?.contains(event.currentTarget)) {
  props.handler(event)
}
```

File: hooks/useClickOutside.ts
Comment (ovflowd) [inline hooks/useClickOutside.ts]: Why mousedown and touchstarts vs click? (Just out of curiosity) (Current approach is valid)

File: hooks/useClickOutside.ts
Comment (AugustinMauroy) [reply hooks/useClickOutside.ts]: why do that ?
what do you thins :
```js
const el = props.ref?.current;
 if (!el || el.contains(event.target as Node)) {
 return;
}
props.handler(event);
```

File: hooks/useClickOutside.ts
Comment (ovflowd) [inline hooks/useClickOutside.ts]:
```suggestion
type TypeHandler = (_event: MouseEvent | TouchEvent) => void;
```

And we need to wait for @HinataKah0 to add the eslint changes.

File: hooks/useClickOutside.ts
Comment (AugustinMauroy) [reply hooks/useClickOutside.ts]: If I don't say any error it's allow to work with mouse and touch

File: hooks/useClickOutside.ts
Comment (shanpriyan) [reply hooks/useClickOutside.ts]: @ovflowd we need to call the handler only if the ref doesn't contain the clicked target element. Also `event.currentTarget` will always point to `document`.


I believe below is the right way to do it
> why do that ?
> what do you thins :
> ```js
> const el = props.ref.current;
>  if (!el || el.contains(event.target as Node)) {
>  return;
> }
> props.handler(event);
> ```

File: hooks/useClickOutside.ts
Comment (shanpriyan) [inline hooks/useClickOutside.ts]: Instead of accepting the ref via props, can we define it inside and return it from the hook

```js
const useClickOutside = ({handler}) => {
const ref = useRef()
//...useEffect
return {ref}
}
```

**Usage** 
```js
const {ref} = useClickOutside({handler: () => setShowDropdown(false)})
return <div ref={ref}>...</div>
```

File: components/Common/LanguageSelector/index.tsx
Comment (shanpriyan) [inline components/Common/LanguageSelector/index.tsx]: since `handler` has been passed in `useClickOutside` hook's `useEffect` deps, handler should be memoized or the useEffect will trigger every re-render
```suggestion
const handler = useCallback(() => setShowDropdown(false),[])

  useClickOutside({
    ref,
    handler
  });
```

File: hooks/useClickOutside.ts
Comment (AugustinMauroy) [reply hooks/useClickOutside.ts:15]: If I understand your comment you say my proposal is good ?

File: hooks/useClickOutside.ts
Comment (AugustinMauroy) [reply hooks/useClickOutside.ts]: I will test that

Comment (AugustinMauroy) [reply hooks/useClickOutside.ts]: It's a good idea but it's create a lot of typing error. So we should keep it as it is

File: hooks/useClickOutside.ts
Comment (ovflowd) [reply hooks/useClickOutside.ts]: Both examples are honestly verbose and unclean. Also please @AugustinMauroy do not resolve the conversation without my ack lol

Comment (ovflowd) [reply hooks/useClickOutside.ts]:
```tsx
if (!props.ref.current?.contains(event.currentTarget)) {
  props.handler(event)
}
```

This should do the trick. We don't need double conditions when we have null coalesce operators. Also we should pretty much use `currentTarget` and NOT `target`

File: hooks/useClickOutside.ts
Comment (ovflowd) [reply hooks/useClickOutside.ts]: You still need to get my question. In some cases, both `mousedown` and `touchstart` will be fired together, but you only want one event.

You shouldn't create event listeners for both events, only one for the corresponding matching event that is appropriate for your device.

```tsx
const clickEvent = 'ontouchstart' in document ? 
  'ontouchstart' : 'onmousedown';

document.addEventListener(clickEvent, listener, { passive: true });

return () => document.removeEventListener(clickEvent, listener);
```

Comment (ovflowd) [reply hooks/useClickOutside.ts]: I'm sure you probably just copied this example on StackOverflow because this code is 99% similar to some StackOverflow examples, which is OK. Still, when copying things from StackOverflow, at least search and try to understand why certain things are done. And if they should be done like that.

File: hooks/useClickOutside.ts
Comment (ovflowd) [reply hooks/useClickOutside.ts]: It's creating typing errors because you probably don't understand how to create types properly.

I agree with @shanpriyan that creating the ref from the Hook is a good idea.

File: hooks/useClickOutside.ts
Comment (AugustinMauroy) [reply hooks/useClickOutside.ts]: my code inspiration form an GitHub repo but but I forgot which one

File: hooks/useClickOutside.ts
Comment (ovflowd) [reply hooks/useClickOutside.ts]: I'm 99% sure the type errors are because of the ambiguity of the destination reference (lack of typing on `useRef` to match any HTMLElement).

This can either be done by providing a low-level element to `useRef<HTMLElement>`, which can probably still give errors, or by creating a generic type:

```tsx
export const useClickOutise = <T extends HTMLElement>(props: UseClickOutsideProps) => {
  const elementRef = useRef<T>();

  // rest of code

  return elementRef;
}

// on the Component
const ref = useClickOutside<HTMLDivElement>(() => {});

return <div ref={ref}>...</div>
```

File: hooks/useClickOutside.ts
Comment (AugustinMauroy) [reply hooks/useClickOutside.ts]: Your proposal work but with `in window`

Comment (ovflowd) [reply hooks/useClickOutside.ts]: StackOverflow / GitHub repo = Doesn't matter 😅

The gist of my message is before simply pasting the code, think about why things are done the way they are and if anything should change. A good Engineer questions code, but also a sound Engineer knows when to listen.

File: components/Common/LanguageSelector/index.tsx
Comment (Harkunwar) [inline components/Common/LanguageSelector/index.tsx]: Since we're passing `() => setShowDropdown(false)` directly from the render function, this will create a new function during each render, causing the `useEffect` inside `useClickOutside` to keep rerunning. You can mitigate the issue by using `useCallback` when you create the render function.

```suggestion
  const dropdownHandler = useCallback(() => setShowDropdown(false), []);
  const ref = useClickOutside<HTMLDivElement>(dropdownHandler);
```

File: hooks/useClickOutside.ts
Comment (Harkunwar) [inline hooks/useClickOutside.ts]: Since `clickEvent` isn't used inside the render function, we should move it inside the `useEffect`

File: hooks/useClickOutside.ts
Comment (Harkunwar) [inline hooks/useClickOutside.ts]: You need to pass `null` to `useRef` for it to infer the correct type, that's why the build is failing. 
```suggestion
  const ref = useRef<T>(null);
```

**Review by Harkunwar — APPROVED** [review]

File: hooks/useClickOutside.ts
Comment (shanpriyan) [inline hooks/useClickOutside.ts]: Just curious, Wouldn't just the `click` event is sufficient enough to make the hook work? 🤔

File: hooks/useClickOutside.ts
Comment (shanpriyan) [inline hooks/useClickOutside.ts]:
```suggestion
      if (!ref.current?.contains(event.target as Node)) {
```
`ref` will never be `null` (so no need of optional chaining there), where as `ref.current` will be `null` until the node gets mounted

File: hooks/useClickOutside.ts
Comment (shanpriyan) [inline hooks/useClickOutside.ts]: nits
```suggestion
    return () => document.removeEventListener(clickEvent, listener);
```

File: hooks/useClickOutside.ts
Comment (shanpriyan) [inline hooks/useClickOutside.ts]: Let's remove this comment after Hinata's PR has merged

**Review by shanpriyan — CHANGES_REQUESTED** [review]

**Comment by shanpriyan** [comment]
@AugustinMauroy please refrain yourself from **resolving** conversations before acknowledging with the reviewer if the review comment is not going to be committed.

File: hooks/useClickOutside.ts
Comment (AugustinMauroy) [reply hooks/useClickOutside.ts]: if I have good memory it's to hollow usage of mouse and tactile device

File: hooks/useClickOutside.ts
Comment (AugustinMauroy) [reply hooks/useClickOutside.ts]: Yeah

File: hooks/useClickOutside.ts
Comment (ovflowd) [inline hooks/useClickOutside.ts]: I still rather not have negated logic, this is over complex.

Comment (AugustinMauroy) [reply hooks/useClickOutside.ts]: I turned the problem in all directions and by spanking the condition with "?" it does not return when you click on the ref

File: hooks/useClickOutside.ts
Comment (HinataKah0) [reply hooks/useClickOutside.ts]: Hmm, I'll cherry-pick the changes to add the rule into another PR tmr so it can be merged first. Because download components migration are blocked by adding the provider first.

File: hooks/useClickOutside.ts
Comment (Harkunwar) [reply hooks/useClickOutside.ts]: The above modification isn't valid because, as the previous condition `el.contains(event.target as Node)` is checking if node exists, your suggestion `!ref.current?.contains(event.target as Node)` checks for if node doesn't exist. 

I would suggest to keep the original `if` condition as it's more simple to understand and read.

File: hooks/useClickOutside.ts
Comment (Harkunwar) [reply hooks/useClickOutside.ts]: I'm actually okay with the way @AugustinMauroy has implemented it.

The reason we can't combine these two conditions is because one is checking for negation, while the other is not checking for that, and both conditions are valid. 

I believe reading it the way it's written right now is much more simple and easy to understand, but combining them with an optional chaining operator makes it more complex to understand what's going on. 

I pointed out the mistake in the previous one here, which is why the tests were failing:
https://github.com/nodejs/nodejs.org/pull/5359#discussion_r1186873691

File: hooks/useClickOutside.ts
Comment (Harkunwar) [inline hooks/useClickOutside.ts:20]:
```suggestion
  }, [ref.current handler]);
```
`ref` will never change so we should change it to `ref.current`

Comment (AugustinMauroy) [reply hooks/useClickOutside.ts:20]: True !

File: hooks/useClickOutside.ts
Comment (Harkunwar) [inline hooks/useClickOutside.ts]:
```suggestion
  }, [ref.current, handler]);
```
Sorry I missed a comma.

Comment (AugustinMauroy) [reply hooks/useClickOutside.ts]: No problem 😁

File: hooks/useClickOutside.ts
Comment (Harkunwar) [reply hooks/useClickOutside.ts]: Maybe we could change it to this to make it more clear:
```typescript
if (ref.current && !ref.current.contains(event.target as Node)) {
  handler(event);
}
```

File: hooks/useClickOutside.ts
Comment (Harkunwar) [inline hooks/useClickOutside.ts]:
```suggestion
      if (ref.current && !ref.current.contains(event.target as Node)) {
        handler(event);
      }
```
Making it more simpler

File: hooks/useClickOutside.ts
Comment (shanpriyan) [reply hooks/useClickOutside.ts]: oops! my bad. Let's stick with the current code 👍🏻

File: hooks/useClickOutside.ts
Comment (shanpriyan) [reply hooks/useClickOutside.ts]: > Maybe we could change it to this to make it more clear:
> 
> ```ts
> if (ref.current && !ref.current.contains(event.target as Node)) {
>   handler(event);
> }
> ```

I would stick with the current code what @AugustinMauroy has raised. This change is just an alternative way of doing that.

File: hooks/useClickOutside.ts
Comment (AugustinMauroy) [reply hooks/useClickOutside.ts]: Done !

File: hooks/useClickOutside.ts
Comment (AugustinMauroy) [reply hooks/useClickOutside.ts]: So I thinks we are good now

File: hooks/useClickOutside.ts
Comment (AugustinMauroy) [reply hooks/useClickOutside.ts]: just click will be ok ?

Comment (shanpriyan) [reply hooks/useClickOutside.ts]: I think so, can you give `click` a try 🤔

File: hooks/useClickOutside.ts
Comment (shanpriyan) [inline hooks/useClickOutside.ts:20]: Change in the `ref.current` property wouldn't trigger the useEffect, I believe just `ref` is enough here.
```suggestion
  }, [ref, handler]);
```

**Review by shanpriyan — APPROVED** [review]

File: hooks/useClickOutside.ts
Comment (shanpriyan) [inline hooks/useClickOutside.ts:20]: Change in the `ref.current` property wouldn't trigger the useEffect, I believe just `ref` is enough here.
```suggestion
  }, [ref, handler]);
```

Comment (shanpriyan) [inline hooks/useClickOutside.ts]: Change in the `ref.current` property wouldn't trigger the useEffect, I believe just `ref` is enough here.

```suggestion
  }, [ref, handler]);
```

**Review by shanpriyan — CHANGES_REQUESTED** [review]

File: hooks/useClickOutside.ts
Comment (AugustinMauroy) [reply hooks/useClickOutside.ts]: I will

Comment (AugustinMauroy) [reply hooks/useClickOutside.ts]: It's working !

File: hooks/useClickOutside.ts
Comment (ovflowd) [inline hooks/useClickOutside.ts:10]:
```suggestion

  useEffect(() => {
```

File: hooks/useClickOutside.ts
Comment (ovflowd) [inline hooks/useClickOutside.ts:19]:
```suggestion

    return () => document.removeEventListener('click', listener);
```

File: hooks/useClickOutside.ts
Comment (ovflowd) [inline hooks/useClickOutside.ts]:
```suggestion
type UseClickOutsideHandler = (_event: MouseEvent | TouchEvent) => void;
```

Comment (ovflowd) [reply hooks/useClickOutside.ts]: This could be fine tuned by doing:

```suggestion
type UseClickOutsideHandler<T extends HTMLElement> = (_event: MouseEvent<T> | TouchEvent<T>) => void;
```

File: hooks/useClickOutside.ts
Comment (ovflowd) [inline hooks/useClickOutside.ts]:
```suggestion
  handler: UseClickOutsideHandler<T>
```

**Review by ovflowd — APPROVED** [review]
LGTM! But I left a few suggestions!

**Comment by shanpriyan** [comment]
LGTM! 🚀 

Good to go after resolving ovflowd's suggestions

**Review by shanpriyan — APPROVED** [review]
LGTM! 🚀

Good to go after resolving ovflowd's suggestions

File: hooks/useClickOutside.ts
Comment (Harkunwar) [reply hooks/useClickOutside.ts:20]: Seems like it's back to ref?
```suggestion
  }, [ref.current, handler]);
```

**Review by aymen94 — APPROVED** [review]

**Pull request merged**
